File: Java/arrayString/DominantIndex.java
package arrayString;

import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

/**
 * In a given integer array nums, there is always exactly one largest element.

Find whether the largest element in the array is AT LEAST TWICE as much as every other number in the array.

If it is, return the index of the largest element, otherwise return -1.
 * */
public class DominantIndex {

	// use too many extra objects => COMPLICATED
	// Find the largest number => Arrays.Sort()
	// Record the index => HashMap
	// Check whether it is at least twice of others => double elements of array, then sort
	// ==> COMPLICATED
	public static int dominantIndexSolution(int[] nums)
	{
		if (nums.length == 0 || nums.length > 50)
        {
            return -1;
        }
		// ONLY ONE ELEMENT
		if (nums.length == 1) return 0;
		Map<Integer, Integer> numMap = new HashMap<Integer, Integer>();
		int[] double2Nums = new int[nums.length];
		
		// find the largest number and its index
		for (int i = 0; i < nums.length; i++)
		{
			numMap.put(i, nums[i]);
			double2Nums[i] = 2 * nums[i];
		}
		Arrays.sort(nums);
		int largest = nums[nums.length - 1];
		Arrays.sort(double2Nums);
		// AT LEAST TWICE as much as every other number
		if (largest >= double2Nums[nums.length - 2])
		{
			int index = numMap.keySet()
			.stream()
			.filter(key -> largest == numMap.get(key))
			.findFirst().get();
			
			return index;
		}
		return -1;
	}
	
	
	// FIND THE LARGEST BY TRAVERSING THE ARRAY
	// CHECK WHETHER IT IS AT LEAST TWICE AS MUCH AS EVERY OTHER NUMBER
	// EXLCUDE ITSELF
	public static int optimizedSolution(int[] nums)
	{
		
		int largest = 0;
		int largestIndex = 0;
		if (nums.length == 0) return -1;
		if (nums.length == 1) return 0; // note the length is 1
		// find the largest number and its index
		for (int i = 0; i < nums.length; i++)
		{
			if (nums[i] > largest)
			{
				largest = nums[i];
				largestIndex = i;
			}
		}
		
		// check whether the largest number is at least twice of others
		// EXCLUDE ITSELF
		for (int i = 0; i < nums.length; i++) {
			if (largest < nums[i] * 2
					&& i != largestIndex) {
				return -1;
			}
		}
		return largestIndex;
	}
	
	
}

File: Python/ArrayAndString/dominant_index.py
# -*- coding: utf-8 -*-
# =============================================================================
# 
# # In a given integer array nums, there is always exactly one largest element.
# #
# # Find whether the largest element in the array is at least twice as much as every other number in the array.
# #
# # If it is, return the index of the largest element, otherwise return -1.
# #
# # Input: nums = [3, 6, 1, 0]
# # Output: 1
# # Explanation: 6 is the largest integer, and for every other number in the array x,
# # 6 is more than twice as big as x.  The index of value 6 is 1, so we return 1.
# #
# # Input: nums = [1, 2, 3, 4]
# # Output: -1
# # Explanation: 4 isn't at least as big as twice the value of 3, so we return -1.
# =============================================================================

def dominantIndex(nums : list) -> int:
    """
    :param nums
    rtype: int
    """
    if not nums: return -1 # empty list
    max_num = max(nums)
    if all(max_num >= x * 2 for x in nums if x != max_num):
        return nums.index(max_num) # get the index of specifized item
    return -1
 

def dominantIndex2(nums: list) -> int:
    if not nums: return -1
    if len(nums) == 1: return 0
    
    max_num = max(nums)
    max_index = nums.index(max_num)
    # remove max number
    nums.remove(max_num)
    
# =============================================================================
#     second_max = max(nums)
#     if max_num >= second_max * 2:
#         return max_index
#     else: pass
#     
#     return -1
# =============================================================================
    for num in nums:
        if 2 * num > max_num:
            return -1
        else: pass
    
    return max_index

if __name__ == '__main__':
    nums = [3, 6, 1, 0]
    # result = dominantIndex(nums)
    result = dominantIndex2(nums)
    print(result)   